panic when deploying a "batch" job.

Given the following job:

```
job "test" {
  datacenters = ["awse"]
  type = "batch"

  constraint {
    attribute = "${meta.role}"
    value     = "utility"
  }
  constraint {
    attribute = "${meta.env}"
    value     = "stag"
  }

  group "test" {

    # set our restart policy
    restart {
      interval = "1m"
      attempts = 2
      delay    = "15s"
      mode     = "fail"
    }
    count = 1

    task "test-batch" {
      artifact {
        source = "<some large file>"
      }
      leader = true
      driver = "exec"
      config {
        command = "bash"
        args = [ "-c", "sleep 30; true" ]
      }
      resources {
        cpu    = 100
        memory = 100
      }
    }

    task "test-service" {
      # grant access to secrets
      driver = "exec"
      config {
        command = "bash"
        args = [ "-c", "sleep infinity" ]
      }
      resources {
        cpu    = 100
        memory = 100
      }
    }
  }
}
```
I will occasionally get the following panic:

```
2018/04/20 17:08:40 EDT [INFO] levant/deploy: triggering a deployment of job test
2018/04/20 17:08:40 EDT [INFO] levant/deploy: evaluation 3ce2139c-0cd0-267d-e9a1-087ce0c30452 finished successfully
2018/04/20 17:08:41 EDT [INFO] levant/job_status_checker: job test has status running
2018/04/20 17:08:46 EDT [INFO] levant/job_status_checker: task test-service in allocation 146785f3-c531-f152-0493-4104f9e66328 has reached running state
2018/04/20 17:08:46 EDT [INFO] levant/job_status_checker: task test-batch in allocation 146785f3-c531-f152-0493-4104f9e66328 has reached running state
panic: assignment to entry in nil map

goroutine 1 [running]:
github.com/jrasell/levant/levant.allocationStatusChecker(0xc4200108a0, 0xc42045ce20, 0x1, 0x4, 0xc420125908)
        /home/dsteen/projects/levant/src/github.com/jrasell/levant/levant/job_status_checker.go:180 +0xe8
github.com/jrasell/levant/levant.(*levantDeployment).jobAllocationChecker(0xc4202ff3f0, 0xc4203baaa0, 0x4)
        /home/dsteen/projects/levant/src/github.com/jrasell/levant/levant/job_status_checker.go:107 +0x14c
github.com/jrasell/levant/levant.(*levantDeployment).jobStatusChecker(0xc4202ff3f0, 0xc4203baaa0, 0x0)
        /home/dsteen/projects/levant/src/github.com/jrasell/levant/levant/job_status_checker.go:34 +0x103
github.com/jrasell/levant/levant.(*levantDeployment).deploy(0xc4202ff3f0, 0xc4202ff3f0)
        /home/dsteen/projects/levant/src/github.com/jrasell/levant/levant/deploy.go:176 +0x9e7
github.com/jrasell/levant/levant.TriggerDeployment(0xc4201ee9c0, 0x2b)
        /home/dsteen/projects/levant/src/github.com/jrasell/levant/levant/deploy.go:66 +0xb7
github.com/jrasell/levant/command.(*DeployCommand).Run(0xc4202b9b60, 0xc42007c080, 0x1, 0x1, 0xc42000cfc0)
        /home/dsteen/projects/levant/src/github.com/jrasell/levant/command/deploy.go:125 +0x4ab
github.com/jrasell/levant/vendor/github.com/mitchellh/cli.(*CLI).Run(0xc420203900, 0xc420203900, 0x4, 0xc42000d040)
        /home/dsteen/projects/levant/src/github.com/jrasell/levant/vendor/github.com/mitchellh/cli/cli.go:255 +0x1eb
main.RunCustom(0xc42007c070, 0x2, 0x2, 0xc4202b99e0, 0xc4202b9920)
        /home/dsteen/projects/levant/src/github.com/jrasell/levant/main.go:49 +0x433
main.Run(0xc42007c070, 0x2, 0x2, 0xb5cdc8)
        /home/dsteen/projects/levant/src/github.com/jrasell/levant/main.go:17 +0x56
main.main()
        /home/dsteen/projects/levant/src/github.com/jrasell/levant/main.go:11 +0x63

```

I only see this about once every 4th run, but the larger the artifact file the more likely it is to occur - which leads me to think this is some sort of race condition.

Thanks!

--- a/levant/job_status_checker.go
+++ b/levant/job_status_checker.go
@@ -10,6 +10,12 @@ import (
 // initially discovered as part of the deployment.
 const initialTaskHealth = "unknown"
 
+// TaskCoordinate is a coordinate for an allocation/task combination
+type TaskCoordinate struct {
+	Alloc    string
+	TaskName string
+}
+
 // jobStatusChecker checks the status of a job at least reaches a status of
 // running. Depending on the type of job and its configuration it can go through
 // more checks.
@@ -77,14 +83,11 @@ func (l *levantDeployment) simpleJobStatusChecker(jobID string) bool {
 // jobs that do not support Nomad deployments.
 func (l *levantDeployment) jobAllocationChecker(evalID *string) bool {
 
-	// Track if we experience any dead tasks in a dumb way.
-	var deadTasks int
-
 	j := l.config.Job.Name
 	q := &nomad.QueryOptions{WaitIndex: 1}
 
 	// Build our small internal checking struct.
-	levantTasks := l.buildAllocationChecker(evalID)
+	levantTasks := make(map[TaskCoordinate]string)
 
 	for {
 
@@ -104,68 +107,17 @@ func (l *levantDeployment) jobAllocationChecker(evalID *string) bool {
 		// If we get here, set the wi to the latest Index.
 		q.WaitIndex = meta.LastIndex
 
-		allocationStatusChecker(levantTasks, allocs, &deadTasks)
+		complete, deadTasks := allocationStatusChecker(levantTasks, allocs)
 
+		// depending on how we finished up we report our status
 		// If we have no allocations left to track then we can exit and log
 		// information depending on the success.
-		if len(levantTasks) == 0 && deadTasks == 0 {
+		if complete && deadTasks == 0 {
 			logging.Info("levant/job_status_checker: all allocations in deployment of job %s are running", *j)
 			return true
-		} else if len(levantTasks) == 0 && deadTasks > 0 {
+		} else if complete && deadTasks > 0 {
 			return false
 		}
-	}
-}
-
-func (l *levantDeployment) buildAllocationChecker(evalID *string) map[string]map[string]string {
-
-	// Create our map to track allocations and tasks within the evaluation created
-	// by the job registration.
-	levantTasks := make(map[string]map[string]string)
-
-	q := &nomad.QueryOptions{WaitIndex: 1}
-
-	// We use a for loop here as, during testing, I have observed Levant runs
-	// faster than Nomad and so without a blocking query a pure GET request can
-	// trigger before Nomad builds the allocation.
-	for {
-
-		// Pull the latest information abou the evaluation allocations from Nomad.
-		allocs, meta, err := l.nomad.Evaluations().Allocations(*evalID, q)
-		if err != nil {
-			logging.Error("levant/job_status_checker: unable to query evaluation for allocations: %v", err)
-			return nil
-		}
-
-		// If the LastIndex is not greater than our stored LastChangeIndex, we don't
-		// need to do anything.
-		if meta.LastIndex <= q.WaitIndex {
-			continue
-		}
-
-		// If we get here, set the wi to the latest Index.
-		q.WaitIndex = meta.LastIndex
-
-		// Iterate over each allocation which can contain multiple tasks in order to
-		// build our object of tasks to check.
-		for _, alloc := range allocs {
-			for taskName := range alloc.TaskStates {
-
-				// If we have not seen the allocation previously we need to init the map.
-				if levantTasks[alloc.ID] == nil {
-					levantTasks[alloc.ID] = make(map[string]string)
-				}
-
-				// Set the task health to our initial status of Unknown.
-				levantTasks[alloc.ID][taskName] = initialTaskHealth
-			}
-		}
-
-		if len(levantTasks) == 0 {
-			continue
-		}
-
-		return levantTasks
 	}
 }
 
@@ -173,37 +125,31 @@ func (l *levantDeployment) buildAllocationChecker(evalID *string) map[string]map
 // job deployment, an update Levants internal tracking on task status based on
 // this. This functionality exists as Nomad does not currently support
 // deployments across all job types.
-func allocationStatusChecker(levantTasks map[string]map[string]string, allocs []*nomad.AllocationListStub, deadTask *int) {
+func allocationStatusChecker(levantTasks map[TaskCoordinate]string, allocs []*nomad.AllocationListStub) (bool, int) {
+
+	complete := true
+	deadTasks := 0
 
 	for _, alloc := range allocs {
 		for taskName, task := range alloc.TaskStates {
-			levantTasks[alloc.ID][taskName] = task.State
-
-			// If the task is running, remove it from tracking.
-			switch levantTasks[alloc.ID][taskName] {
-			case nomadStructs.TaskStateRunning:
-				logging.Info("levant/job_status_checker: task %s in allocation %s has reached %s state",
-					taskName, alloc.ID, nomadStructs.TaskStateRunning)
-				delete(levantTasks[alloc.ID], taskName)
-
-			case nomadStructs.TaskStatePending:
-				logging.Debug("levant/job_status_checker: task %s in allocation %s now in %s state",
-					taskName, alloc.ID, nomadStructs.TaskStatePending)
-
-			// If the task is dead, incrament the deadTask counter and remove the task
-			// from tracking.
-			case nomadStructs.TaskStateDead:
-				logging.Error("levant/job_status_checker: task %s in allocation %s now in %s state",
-					taskName, alloc.ID, nomadStructs.TaskStateDead)
-				*deadTask++
-				delete(levantTasks[alloc.ID], taskName)
+			// if the state is one we haven't seen yet then we print a message
+			if levantTasks[TaskCoordinate{alloc.ID, taskName}] != task.State {
+				logging.Info("levant/job_status_checker: task %s in allocation %s now in %s state",
+					taskName, alloc.ID, task.State)
+				// then we record the new state
+				levantTasks[TaskCoordinate{alloc.ID, taskName}] = task.State
 			}
 
-			// If we have no tasks left under the allocation to track remove the
-			// allocation from our tracker.
-			if len(levantTasks[alloc.ID]) == 0 {
-				delete(levantTasks, alloc.ID)
+			// then we have some case specific actions
+			switch levantTasks[TaskCoordinate{alloc.ID, taskName}] {
+			// if a task is still pendign we are not yet done
+			case nomadStructs.TaskStatePending:
+				complete = false
+				// if the task is dead we record that
+			case nomadStructs.TaskStateDead:
+				deadTasks++
 			}
 		}
 	}
+	return complete, deadTasks
 }

--- a/levant/job_status_checker_test.go
+++ b/levant/job_status_checker_test.go
@@ -9,19 +9,10 @@ import (
 
 func TestJobStatusChecker_allocationStatusChecker(t *testing.T) {
 
-	// Setup our LevantTask structures.
-	levantTasks1 := make(map[string]map[string]string)
-	levantTasks1["10246d87-ecd7-21ad-13b2-f0c564647d64"] = make(map[string]string)
-	levantTasks1["10246d87-ecd7-21ad-13b2-f0c564647d64"]["task1"] = initialTaskHealth
-
-	levantTasks2 := make(map[string]map[string]string)
-	levantTasks2["20246d87-ecd7-21ad-13b2-f0c564647d64"] = make(map[string]string)
-	levantTasks2["20246d87-ecd7-21ad-13b2-f0c564647d64"]["task1"] = initialTaskHealth
-	levantTasks2["20246d87-ecd7-21ad-13b2-f0c564647d64"]["task2"] = initialTaskHealth
-
-	levantTasks3 := make(map[string]map[string]string)
-	levantTasks3["30246d87-ecd7-21ad-13b2-f0c564647d64"] = make(map[string]string)
-	levantTasks3["30246d87-ecd7-21ad-13b2-f0c564647d64"]["task1"] = initialTaskHealth
+	// Build our task status maps
+	levantTasks1 := make(map[TaskCoordinate]string)
+	levantTasks2 := make(map[TaskCoordinate]string)
+	levantTasks3 := make(map[TaskCoordinate]string)
 
 	// Build a small AllocationListStubs with required information.
 	var allocs1 []*nomad.AllocationListStub
@@ -50,43 +41,43 @@ func TestJobStatusChecker_allocationStatusChecker(t *testing.T) {
 	})
 
 	cases := []struct {
-		levantTasks    map[string]map[string]string
-		allocs         []*nomad.AllocationListStub
-		dead           int
-		expectedDead   int
-		expectedAllocs int
+		levantTasks      map[TaskCoordinate]string
+		allocs           []*nomad.AllocationListStub
+		dead             int
+		expectedDead     int
+		expectedComplete bool
 	}{
 		{
 			levantTasks1,
 			allocs1,
 			0,
 			0,
-			0,
+			true,
 		},
 		{
 			levantTasks2,
 			allocs2,
 			0,
 			0,
-			1,
+			false,
 		},
 		{
 			levantTasks3,
 			allocs3,
 			0,
 			1,
-			0,
+			true,
 		},
 	}
 
 	for _, tc := range cases {
-		allocationStatusChecker(tc.levantTasks, tc.allocs, &tc.dead)
+		complete, dead := allocationStatusChecker(tc.levantTasks, tc.allocs)
 
-		if len(tc.levantTasks) != tc.expectedAllocs {
-			t.Fatalf("expected %v but got %v", tc.expectedAllocs, len(tc.levantTasks))
+		if complete != tc.expectedComplete {
+			t.Fatalf("expected complete to be %v but got %v", tc.expectedComplete, complete)
 		}
-		if tc.dead != tc.expectedDead {
-			t.Fatalf("expected %v dead task(s) but got %v", tc.expectedDead, tc.dead)
+		if dead != tc.expectedDead {
+			t.Fatalf("expected %v dead task(s) but got %v", tc.expectedDead, dead)
 		}
 	}
 }

